### PR TITLE
Implement device node scanning for Birger lens adaptors (#44)

### PR DESCRIPTION
**Pull request by AnthonyHorton**
Quick implementation done in the blind, but I think it works.

This will allow Birger lens adaptors to be specified in the `pocs.yaml` with either a device node (same as before) or a serial number in the `port` field.  If a 5 digit number is given it is assumed to be a serial number and the Birger focuser constructor will step through all device nodes matching a certain pattern (default `/dev/tty.USA49GW*.?` to match Tripplite Keyspan USA-49GW USB-serial convertor nodes), opening each in turn and sending Birger commands to attempt to retrieve the serial number until a match is found.

This will add to startup time but ensures that the right Birger adaptor gets associated with the right camera, even when device nodes change.

### CODE REVIEW

**Comment by AnthonyHorton** [comment]
This could be made more efficient. As written this will do the device node scan for each Birger adaptor that's been specified by serial number. Ideally the scanning should be done once when the 1st Birger object is created, and the results cached for the use by subsequent Birgers.

This would also avoid the device node scanning process talking to *all* the Birgers, including those used by other cameras.

**Comment by AnthonyHorton** [comment]
Rewrote this to do the above, i.e. scan device nodes once and once only the first time that a Birger object is created using a serial number, and cache the results in a class variable.  Also tracks which device nodes have already been claimed in order to avoid scanning those nodes, and to check for conflicts.

File: pocs/focuser/birger.py [hunk 1]
```diff
@@ -2,24 +2,69 @@
...
+                        self._birger_nodes[serial_number] = device_node
+                    except serial.SerialException, serial.SerialTimeoutException, AssertionError:
+                        # No birger on this node.
+                        pass
+                    finally:
+                        self._serial_port.close()
+
+            # Search in cached device node scanning results for serial number
+            try:
+                device_node = self._birger_nodes[self.port]
```
Comment (lspitler) [inline pocs/focuser/birger.py]: Back on line 45 you use the key `serial_number` on dictionary(?) `self._birger_nodes`. Here you use `self.port` key. Is this correct?

File: pocs/focuser/birger.py [hunk 2]
```diff
@@ -2,24 +2,69 @@
...
+            except KeyError:
+                self.logger.critical("Could not find {} ({})!".format(self.name, self.port))
+                return
+            self.port = device_node
```
Comment (lspitler) [inline pocs/focuser/birger.py]: Mmm, line 45 assigns `device_node` to `self._birger_nodes`, which later is referenced by `self.port` on line 54, which here is being set to `device_node`. If this is all correct it is a little hard to parse reading from top to bottom, but fixing it to be more readable is probably less important than it working now.

File: pocs/camera/camera.py [hunk 3]
```diff
@@ -213,8 +213,8 @@ def autofocus(self,
 
     def get_thumbnail(self, seconds, file_path, thumbnail_size, keep_files=False):
         """
```
Comment (wtgee) [inline pocs/camera/camera.py]: Can you actually do a short one line statement for first sentence and then more desc after?

```
"""Gets a thumbnail from centre of image

Takes an images, grabs the data, deletes...
```

That's the encouraged form as code hints in editors will just try to read that very top line.

File: pocs/focuser/birger.py [hunk 4]
```diff
@@ -2,24 +2,69 @@
...
                  name='Birger Focuser',
                  model='Canon EF-232',
                  initial_position=None,
+                 dev_node_pattern='/dev/tty.USA49WG*.?',
                  *args, **kwargs):
         super().__init__(name=name, model=model, *args, **kwargs)
         self.logger.debug('Initialising Birger focuser')
-        self.connect()
+
+        if serial_number_pattern.match(self.port):
```
Comment (wtgee) [inline pocs/focuser/birger.py]: Where does this come from?

Comment (wtgee) [reply pocs/focuser/birger.py]: I see it at bottom of file. See comment there.

File: pocs/focuser/birger.py [hunk 5]
```diff
@@ -106,7 +151,7 @@ def connect(self):
 
         except serial.SerialException as err:
             self._serial_port = None
-            self.logger.critical('Could not connect to {}!'.format(self))
+            self.logger.critical('Could not open {}!'.format(port))
             raise err
 
         time.sleep(2)
```
Comment (wtgee) [inline pocs/focuser/birger.py]: I had this in some other `SerialData` class and was questioned about why it was there and didn't have an answer. Is this recommended? Either way we probably want a comment and maybe want to get rid of magic number.

File: pocs/focuser/birger.py [hunk 6]
```diff
@@ -316,6 +365,9 @@ def _move_inf(self):
             return int(r[:-2])
 
 
+# Birger adaptor serial numbers should be 5 digits
+serial_number_pattern = re.compile('^\d{5}$')
```
Comment (wtgee) [inline pocs/focuser/birger.py]: Can we move these to the top of file?

Comment (AnthonyHorton) [reply pocs/focuser/birger.py]: OK.

File: pocs/focuser/birger.py (hunk 5 above)
Comment (AnthonyHorton) [reply pocs/focuser/birger.py]: I don't remember why it's there. Can try removing it/changing the number once we can test with the real hardware.

File: pocs/camera/camera.py (hunk 3 above)
Comment (AnthonyHorton) [reply pocs/camera/camera.py]: Yep.

File: pocs/focuser/birger.py (hunk 1 above)
Comment (AnthonyHorton) [reply pocs/focuser/birger.py]: It's correct, albeit confusing, for historical reasons.

File: pocs/focuser/birger.py (hunk 2 above)
Comment (AnthonyHorton) [reply pocs/focuser/birger.py]: Again, I think it is correct, but it is somewhat confusing.

Originally `self.port` (and the corresponding `port:` entry in the YAML file) was used by the PANOPTES DSLR cameras to specify their device node. When I added the SBIG cameras I continued to use `port` to specify the device as I was inheriting from the same base class, but because the SBIG driver doesn't work with device nodes I stretched the meaning to include serial numbers, that's how the SBIG cameras are specified in the YAML file.  When I later added the Birger focuser I used `port` again, once more to specify a device node.  Now I've added the option of specifying the Birgers by serial number instead, using the same `port:` entry in the YAML file. So, in a case where you do specify a serial number `self.port` starts off containing a serial number, but later gets replaced by the corresponding device node. What I should probably do is use separate, `serial_number` or `port` entries rather than misusing `port` in the SBIG and Birger classes.
